tests/test_segment_producer.py

```diff
@@ -0,0 +1,197 @@
+)
+
+data_info = collections.namedtuple("data_info", ["data", "md5sum"])
+
+
+@pytest.fixture()
+def setup_directories(tmp_path: pathlib.Path):
+    data_directory = tmp_path.joinpath("test")
+    state_directory = tmp_path.joinpath("test/logs")
+    tmp_path.mkdir(exist_ok=True)
+    data_directory.mkdir(exist_ok=True)
+    state_directory.mkdir(exist_ok=True)
+
+    return directories_tuple(tmp_path, data_directory, state_directory)
+
+
+@pytest.fixture()
+def complete_data():
+    data = b"A" * 1024
+    return data_info(data, utils.md5sum(data))
+
+
+@pytest.fixture()
+def incomplete_data():
+    data = b"A" * 512
+    return data_info(data, utils.md5sum(data))
+
+
+@pytest.fixture()
+def mock_download_stream(monkeypatch, setup_directories: typing.NamedTuple):
+    def mock_init(self):
+        self.initialized = True
+        self.name = "test.txt"
+        self.size = 1024
+        self.md5sum = "d47b127bc2de2d687ddc82dac354c415"
+
+    monkeypatch.setattr(stream.DownloadStream, "init", mock_init)
+
+    download_stream = stream.DownloadStream(
+        url="https://localhost:80/data/test",
+        directory=str(setup_directories.base_directory.resolve()),
+        token=None,
+    )
+    download_stream.init()
+
+    return download_stream
+
+
+@pytest.fixture(autouse=True)
+def mock_schedule(monkeypatch):
+    """Mock SegmentProducer.schedule() to not immediately fill up work queue with Intervals"""
+
+    def mock_schedule(self):
+        return
+
+    monkeypatch.setattr(segment.SegmentProducer, "schedule", mock_schedule)
+
+
+@pytest.fixture()
+def mock_complete_download_file(
+    setup_directories: typing.NamedTuple, complete_data: typing.NamedTuple
+):
+    write_data_file(setup_directories.data_directory, "test.txt", complete_data.data)
+
+
+@pytest.fixture()
+def mock_incomplete_download_file(
+    setup_directories: typing.NamedTuple, incomplete_data: typing.NamedTuple
+):
+    write_data_file(setup_directories.data_directory, "test.txt", incomplete_data.data)
+
+
+@pytest.fixture()
+def mock_temporary_file(
+    setup_directories: typing.NamedTuple, incomplete_data: typing.NamedTuple
+):
+    write_data_file(
+        setup_directories.data_directory, "test.txt.partial", incomplete_data.data
+    )
+
+
+@pytest.fixture()
+def mock_complete_state_file(
+    setup_directories: typing.NamedTuple, complete_data: typing.NamedTuple
+):
+    write_state_file(
+        setup_directories.state_directory, "test.txt.parcel", complete_data
+    )
+
+
+@pytest.fixture()
+def mock_incomplete_state_file(
+    setup_directories: typing.NamedTuple, incomplete_data: typing.NamedTuple
+):
+    write_state_file(
+        setup_directories.state_directory, "test.txt.parcel", incomplete_data
+    )
+
+
+def write_data_file(directory: pathlib.Path, file_name: str, data: bytes):
+    file_path = directory.joinpath(file_name)
+    with file_path.open("wb") as f:
+        f.write(data)
+
```

Comment:
Do any of the consumers of this function clean up the files when they're done?

Reply:
no, from my understanding `tmp_path` creates a new path (which is the base directory where all the created files fall under) for every test run, and pytest will keep the latest 3 `tmp_path` directories (useful for checking file state after a failed test run). So the latest three `tmp_path` directories will remain. If we want to explicitly delete the base directory after each test run we can, its just if you have a failed test for some reason, you won't be able to go inspect the files afterwards (which I ran into).

Reply:
Oh I see. I misread the `setup_directories` fixture. Type annotations in fixtures trip me up sometimes. I thought `tmp_path` was the parameter and not another fixture.

Reply:
ah, thats a good point. should we not be type hinting fixtures? looking at it now it is kind of confusing